iModelJsNodeAddon/JsLogger.cpp

```diff
@@ -115,15 +126,27 @@ void JsLogger::LogMessage(Utf8CP category, SEVERITY sev, Utf8CP msg) {
 
     if (canUseJavaScript()) {
         processDeferred();
-        logToJs(category, sev, msg);
-    } else {
-        // save this message in memory so it can be logged later on the JavaScript thread
-        m_deferredLogging.push_back(LoggedMessage(category, sev, msg));
+        try {
+            logToJs(category, sev, msg);
+            return;
+        } catch (...) {
+            // Push to deferred below, because JS log failed.
+            // logToJs triggers the backend JavaScript logging functionality. That functionality
+            // allows for custom log functions. Those custom log functions may throw an exception,
+            // and if they do so, that exception gets propogated to here. Catch that here so that
+            // we don't throw a C++ exception from this function.
+        }
     }
+    // save this message in memory so it can be logged later on the JavaScript thread
+    m_deferredLogging.push_back(LoggedMessage(category, sev, msg));
 }
 
 bool JsLogger::IsSeverityEnabled(Utf8CP category, SEVERITY sev) {
     BeMutexHolder lock(m_deferredLogMutex);
+    if (canUseJavaScript()) {

```

Comment:
please remove this change. Why did you add it?

Reply:
I added it because deferred error messages were basically never getting sent, since the only way for them to be sent was for another _error_ to be logged. With this change, they get sent the next time any severity is logged (even ones that are disabled). I can remove the change if you really want, but I feel that it is an improvement over what was there before.

Reply:
Yes, _please_ remove this. `IsSeverityEnabled` is called way more often than actual logging (in fact, it's often called multiple times for the same log site) and should not have any side effects. Differed messages are deferred until another log request (any log requests, not errors) happens on the main thread. That is not an urgent situation.

Reply:
Yes, log messages are deferred until another **enabled** message is sent on the main thread - that doesn't have to be an error. It is true that it's indeterminate when that may happen. That's why the `flushLog` method exists.

Reply:
no, it can only be called from JavaScript, so `canUseJavaScript` is not necessary.